File: android/app/src/main/java/com/noop/ui/SleepLayoutPrefs.kt
package com.noop.ui

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.ui.res.stringResource
import com.noop.R

// MARK: - Reorderable Sleep sections (#sleep-layout)
//
// The Sleep tab's analytical cards — Sleep marks, the Stages hypnogram, Naps, Night detail, the Sleep-debt
// ledger, Stages-vs-typical, and the Asleep-duration trend — render in one fixed order below the pinned
// Sleep-performance hero + date navigator. This lets the user REORDER or HIDE those cards, mirroring the
// Today tab's Arrange sheet (see TodayLayoutPrefs), with the default being the original order so nothing
// changes for anyone who never customizes Sleep. Display-only — no metric is computed or stored
// differently; this only decides which already-built cards render and in what sequence.
//
// Stored as a single comma-joined string of section keys in SharedPreferences ("sleep.sectionOrder"), the
// same mechanism TodayLayoutPrefs/KeyMetricPrefs use. Mirrors the macOS SleepLayoutPrefs.swift +
// @AppStorage("sleep.sectionOrder"). Every known section stays in the order registry: unknown tokens are
// dropped, and missing known sections are inserted at their default positions. Explicit reversible
// visibility lives separately in "sleep.hiddenSections", byte-identical to the Apple-platform key.

/**
 * One reorderable Sleep card. The [raw] is the stable persisted identifier — keep it byte-identical to the
 * macOS `SleepSection` enum so a backup/restore reads the same layout on either OS.
 */
enum class SleepSection(val raw: String, val title: String) {
    SLEEP_MARKS("sleepMarks", "Sleep marks"),
    STAGES("stages", "Stages"),
    NIGHT_DETAIL("nightDetail", "Night detail"),
    SLEEP_DEBT("sleepDebt", "Sleep-debt ledger"),
    STAGES_VS_TYPICAL("stagesVsTypical", "Stages vs typical"),
    ASLEEP_DURATION("asleepDuration", "Asleep duration"),

    /** #sleep-layout: two ANDROID-ONLY detail cards (Hours-vs-Needed + Consistency, richer than the
     *  Night-detail grid tiles) — previously pinned below the arrange region, now first-class arrangeable
     *  Sleep sections. iOS deliberately renders these metrics only as Night-detail grid tiles (tap-through),
     *  so these two rawValues are Android-only — the macOS `SleepSection` stops at `asleepDuration`. Safe to
     *  diverge here: `sleep.sectionOrder` is not in the .noopbak whitelist, so a cross-OS restore never
     *  reads them. (Consistency's underlying score also differs across platforms — a separate parity item.) */
    HOURS_VS_NEEDED("hoursVsNeeded", "Hours vs Needed"),
    CONSISTENCY("consistency", "Consistency");

    companion object {
        fun fromRaw(raw: String?): SleepSection? = entries.firstOrNull { it.raw == raw }

        /** The original, hard-coded card order — the default when the layout isn't customised. Matches the
         *  pre-customisation render order in SleepScreen below the pinned Sleep-performance hero. (Naps
         *  rides with Stages for now — drawn inside the stages hero; making it an independently arrangeable
         *  card is a follow-up that requires hoisting the hero's edit/delete callbacks.) */
        val defaultOrder: List<SleepSection> = listOf(
            SLEEP_MARKS, STAGES, NIGHT_DETAIL, SLEEP_DEBT, STAGES_VS_TYPICAL, ASLEEP_DURATION,
            HOURS_VS_NEEDED, CONSISTENCY,
        )
    }
}

/**
 * The section's display title, localized. The enum's [title] field stays the English source-of-truth
 * default (used for logging/comparisons); the UI reads this so the Sleep arrange sheet shows a translated
 * title. Enum constructors can't call [stringResource], so resolution happens here at the render site.
 * Mirrors the Apple platforms, where `SleepSection.title` is a `String(localized:)`.
 */
@Composable
fun SleepSection.localizedTitle(): String = when (this) {
    SleepSection.SLEEP_MARKS -> stringResource(R.string.l10n_sleep_screen_sleep_marks_8e9b86f0)
    SleepSection.STAGES -> stringResource(R.string.l10n_sleep_screen_stages_c1d33ad5)
    SleepSection.NIGHT_DETAIL -> stringResource(R.string.l10n_sleep_screen_night_detail_8f271bcf)
    SleepSection.SLEEP_DEBT -> stringResource(R.string.l10n_sleep_screen_sleep_debt_ledger_8cc9a992)
    SleepSection.STAGES_VS_TYPICAL -> stringResource(R.string.l10n_sleep_screen_stages_vs_typical_28463f24)
    SleepSection.ASLEEP_DURATION -> stringResource(R.string.l10n_sleep_screen_asleep_duration_3638413f)
    SleepSection.HOURS_VS_NEEDED -> stringResource(R.string.l10n_sleep_screen_hours_vs_needed_500a0aca)
    SleepSection.CONSISTENCY -> stringResource(R.string.l10n_sleep_screen_consistency_0ea7b95e)
}

/**
 * Display-only persistence for the Sleep card order and explicit hidden set. Every known card remains in
 * the order registry, while visibility is stored separately so hiding stays reversible. SharedPreferences
 * isn't reactive, so Sleep reads this into remembered state and refreshes it after save. Mirrors the
 * Apple-platform SleepLayoutPrefs, a direct twin of TodayLayoutPrefs with a `sleep.` key namespace.
 */
object SleepLayoutPrefs {
    private const val KEY_ORDER = "sleep.sectionOrder"
    private const val KEY_HIDDEN = "sleep.hiddenSections"

    /** Every known card in display order (saved order first, then any newly-added card at its default position). */
    fun order(context: Context): List<SleepSection> =
        decodeOrder(NoopPrefs.of(context).getString(KEY_ORDER, null))

    /** Persist the card order. */
    fun setOrder(context: Context, sections: List<SleepSection>) {
        NoopPrefs.of(context).edit().putString(KEY_ORDER, encode(sections)).apply()
    }

    /** The explicitly hidden cards in their editor order. Empty/unset means every card is visible. */
    fun hidden(context: Context): List<SleepSection> =
        decodeHidden(NoopPrefs.of(context).getString(KEY_HIDDEN, null))

    /** Persist the explicit reversible hidden list. */
    fun setHidden(context: Context, sections: List<SleepSection>) {
        NoopPrefs.of(context).edit().putString(KEY_HIDDEN, encodeHidden(sections)).apply()
    }

    /** Encode an ordered list of cards into the stored comma-joined string. */
    fun encode(sections: List<SleepSection>): String = sections.joinToString(",") { it.raw }

    fun encodeHidden(sections: List<SleepSection>): String = sections.joinToString(",") { it.raw }

    /**
     * Decode the stored string into the FULL ordered card list. An empty/unset string yields the default
     * order. Unknown tokens are ignored, duplicates collapsed, and any known card missing from the saved
     * order is INSERTED at its default-order position relative to the saved cards (before the first saved
     * card that follows it in the default order; appended when none does) — so every card always renders,
     * and one added in a later app version surfaces where users expect it instead of teleporting to the
     * bottom of an existing saved order.
     */
    fun decodeOrder(raw: String?): List<SleepSection> {
        val trimmed = raw?.trim().orEmpty()
        if (trimmed.isEmpty()) return SleepSection.defaultOrder
        val saved = ArrayList<SleepSection>()
        trimmed.split(",").forEach { token ->
            SleepSection.fromRaw(token.trim())?.let { if (it !in saved) saved.add(it) }
        }
        if (saved.isEmpty()) return SleepSection.defaultOrder
        // Iterate entries (not defaultOrder) so a future enum case accidentally left out of defaultOrder
        // can never be silently hidden; a card without a default index sorts after everything. Twin of the
        // Swift decodeOrder's degraded path; defaultOrder covering every entry is pinned by the tests.
        fun defIdx(s: SleepSection): Int =
            SleepSection.defaultOrder.indexOf(s).let { if (it == -1) SleepSection.defaultOrder.size else it }
        for (missing in SleepSection.entries) {
            if (missing in saved) continue
            val insertAt = saved.indexOfFirst { defIdx(it) > defIdx(missing) }
            if (insertAt == -1) saved.add(missing) else saved.add(insertAt, missing)
        }
        return saved
    }

    /**
     * Decode only explicitly hidden cards. Unknown tokens are ignored and duplicates collapsed. Missing
     * entries stay visible, so a card added by a future release automatically surfaces.
     */
    fun decodeHidden(raw: String?): List<SleepSection> {
        val hidden = LinkedHashSet<SleepSection>()
        raw?.split(",")?.forEach { token -> SleepSection.fromRaw(token.trim())?.let(hidden::add) }
        return hidden.toList()
    }

    /** The full saved order filtered by the explicit hidden set. */
    fun visibleOrder(orderRaw: String?, hiddenRaw: String?): List<SleepSection> {
        val hidden = decodeHidden(hiddenRaw).toSet()
        return decodeOrder(orderRaw).filterNot { it in hidden }
    }
}

File: Strand/Data/SleepLayoutPrefs.swift
import Foundation
import SwiftUI

// MARK: - Reorderable Sleep sections (#sleep-layout)
//
// The Sleep tab's analytical cards — Sleep marks, the Stages hypnogram, Naps, Night detail, the Sleep-debt
// ledger, Stages-vs-typical, and the Asleep-duration trend — render in one fixed order below the pinned
// Sleep-performance hero + date navigator. This lets the user REORDER or HIDE those cards, mirroring the
// Today tab's Arrange sheet (see `TodayLayoutPrefs`), with the default being the original order so nothing
// changes for anyone who never customises Sleep. Display-only — no metric is computed or stored
// differently; this only decides which already-built cards render and in what sequence.
//
// Stored as a single comma-joined string of section keys in @AppStorage("sleep.sectionOrder"), the same
// mechanism `TodayLayoutPrefs`/`KeyMetricPrefs` use. The Android side mirrors this byte-identically in
// SleepLayoutPrefs.kt (SharedPreferences "sleep.sectionOrder"). Every known section stays in the ORDER
// registry: unknown tokens are dropped, and any known section missing from the saved order is INSERTED at
// its default-order position relative to the saved sections — so a card added in a later version surfaces
// where users expect it rather than teleporting to the bottom of an existing saved order.
//
// User visibility is stored separately in "sleep.hiddenSections". Keeping order and visibility separate is
// intentional: hiding is reversible, a hidden card keeps its stable identity, and a card introduced by a
// future version defaults to visible because it is absent from the explicit hidden set.
//
// The Sleep-performance hero and the date navigator are NOT reorderable — they are the fixed frame of the
// tab, exactly as the Today hero is pinned above its arrangeable sections.

/// One reorderable Sleep card. The rawValue is the stable persisted identifier — keep it byte-identical to
/// the Android `SleepSection` enum so a backup/restore reads the same layout on either OS.
enum SleepSection: String, CaseIterable, Identifiable {
    case sleepMarks
    case stages
    case nightDetail
    case sleepDebt
    case stagesVsTypical
    case asleepDuration

    var id: String { rawValue }

    /// The card's display label in the Arrange sheet — matches the Android `SleepSection.title`.
    var title: String {
        switch self {
        case .sleepMarks:      return String(localized: "Sleep marks")
        case .stages:          return String(localized: "Stages")
        case .nightDetail:     return String(localized: "Night detail")
        case .sleepDebt:       return String(localized: "Sleep-debt ledger")
        case .stagesVsTypical: return String(localized: "Stages vs typical")
        case .asleepDuration:  return String(localized: "Asleep duration")
        }
    }

    /// The original, hard-coded card order — the default when the layout isn't customised. Matches the
    /// pre-customisation render order in `SleepView` below the pinned Sleep-performance hero. (Naps rides
    /// with Stages for now — it's drawn inside the stages hero; making it an independently arrangeable
    /// card is a follow-up that requires hoisting the hero's edit/delete callbacks.)
    static let defaultOrder: [SleepSection] = [
        .sleepMarks, .stages, .nightDetail, .sleepDebt, .stagesVsTypical, .asleepDuration,
    ]
}

/// Display-only persistence for the Sleep card order and visibility. The order registry always contains
/// every known card; `hiddenKey` stores the explicit reversible hidden set. Mirrors Android byte-for-byte,
/// and is a direct twin of `TodayLayoutPrefs` with a `sleep.` key namespace.
enum SleepLayoutPrefs {
    /// UserDefaults key — a comma-joined list of `SleepSection` rawValues in display order.
    static let orderKey = "sleep.sectionOrder"
    /// UserDefaults key — a comma-joined list of explicitly hidden `SleepSection` rawValues.
    static let hiddenKey = "sleep.hiddenSections"

    /// Encode an ordered card list into the stored comma-joined string.
    static func encode(_ sections: [SleepSection]) -> String {
        sections.map(\.rawValue).joined(separator: ",")
    }

    /// Encode the explicit hidden set in stable list order. The editor passes its Hidden-card order;
    /// rendering treats the decoded value as a set.
    static func encodeHidden(_ sections: [SleepSection]) -> String {
        sections.map(\.rawValue).joined(separator: ",")
    }

    /// Decode the stored string into the FULL ordered card list. An empty/unset string yields the default
    /// order. Unknown tokens are ignored, duplicates collapsed, and any known card missing from the saved
    /// order is INSERTED at its default-order position relative to the saved cards (before the first saved
    /// card that follows it in the default order; appended when none does) — so every card always renders,
    /// and one added in a later app version surfaces where users expect it instead of teleporting to the
    /// bottom of an existing saved order. Twin of the Kotlin `decodeOrder`.
    static func decodeOrder(_ raw: String) -> [SleepSection] {
        let trimmed = raw.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return SleepSection.defaultOrder }
        var saved: [SleepSection] = []
        for token in trimmed.split(separator: ",") {
            if let s = SleepSection(rawValue: token.trimmingCharacters(in: .whitespaces)), !saved.contains(s) {
                saved.append(s)
            }
        }
        guard !saved.isEmpty else { return SleepSection.defaultOrder }
        // Iterate allCases (not defaultOrder) so a future case accidentally left out of defaultOrder can
        // never be silently hidden; a card without a default index sorts after everything. defaultOrder
        // covering allCases is pinned by SleepLayoutPrefsTests on both platforms.
        func defIdx(_ s: SleepSection) -> Int {
            SleepSection.defaultOrder.firstIndex(of: s) ?? SleepSection.defaultOrder.count
        }
        for missing in SleepSection.allCases where !saved.contains(missing) {
            let insertAt = saved.firstIndex { defIdx($0) > defIdx(missing) }
            if let insertAt { saved.insert(missing, at: insertAt) } else { saved.append(missing) }
        }
        return saved
    }

    /// Decode explicitly hidden cards. Empty/unset means nothing is hidden. Unknown tokens are ignored and
    /// duplicates collapsed; unlike `decodeOrder`, missing cases are NOT inserted because absence here means
    /// visible (including a card introduced by a future app version).
    static func decodeHidden(_ raw: String) -> [SleepSection] {
        var seen = Set<SleepSection>()
        var hidden: [SleepSection] = []
        for token in raw.split(separator: ",") {
            if let section = SleepSection(rawValue: token.trimmingCharacters(in: .whitespaces)),
               seen.insert(section).inserted {
                hidden.append(section)
            }
        }
        return hidden
    }

    /// The cards Sleep should render, preserving the full saved order while filtering only the user's
    /// explicit hidden set. At least one visible card is enforced by the editor, not the decoder.
    static func visibleOrder(orderRaw: String, hiddenRaw: String) -> [SleepSection] {
        let hidden = Set(decodeHidden(hiddenRaw))
        return decodeOrder(orderRaw).filter { !hidden.contains($0) }
    }
}
